packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts

```diff
@@ -331,6 +332,69 @@ export const useGridRows = (
     [apiRef, logger],
   );
 
+  const replaceRows = React.useCallback<GridRowApi['replaceRows']>(
+    (firstRowToRender, lastRowToRender, newRows) => {
+      if (props.signature === GridSignature.DataGrid) {
+        throw new Error(
+          [
+            "MUI: You can't replace rows using `apiRef.current.replaceRows` on the DataGrid.",
+            'You need to upgrade to DataGridPro or DataGridPremium component to unlock this feature.',
+          ].join('\n'),
+        );
+      }
+
+      if (newRows.length && newRows.length !== lastRowToRender - firstRowToRender) {
+        throw new Error(
+          [
+            'MUI: The number of rows you want update is different than what was indicated',
+            'You need to pass rows which number is equal to the difference between the `firstRowToRender` and `lastRowToRender`.',
+          ].join('\n'),
+        );
+      }
+
+      const allRows = gridRowIdsSelector(apiRef);
+      const updatedRows = [...allRows];
+      const idRowsLookup = gridRowsLookupSelector(apiRef);
+      const idToIdLookup = gridRowsIdToIdLookupSelector(apiRef);
+      const tree = gridRowTreeSelector(apiRef);
+      const updatedIdRowsLookup = { ...idRowsLookup };
+      const updatedIdToIdLookup = { ...idToIdLookup };
+      const updatedTree = { ...tree };
+
+      newRows.forEach((row, index) => {
+        const rowId = getRowIdFromRowModel(
+          row,
+          props.getRowId,
+          'A row was provided without id when calling replaceRows().',
+        );
+        const [replacedRowId] = updatedRows.splice(firstRowToRender + index, 1, rowId);
+
+        delete updatedIdRowsLookup[replacedRowId];
+        delete updatedIdToIdLookup[replacedRowId];
+        delete updatedTree[replacedRowId];
+      });
+
+      newRows.forEach((row) => {
+        updatedIdRowsLookup[row.id] = row;
+        updatedIdToIdLookup[row.id] = row.id;
+        updatedTree[row.id] = row as GridRowTreeNodeConfig;
```

Comment:
This is what I mentioned in https://github.com/mui/mui-x/pull/5214#discussion_r939047535
updatedTree should contain `GridRowTreeNodeConfig`, not `GridRowModel`